gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java

```diff
@@ -69,12 +69,34 @@ public void write(JsonWriter out, T value) throws IOException {
     chosen.write(out, value);
   }
 
+  /**
+   * @param typeAdapter the type adapter to check.
+   * @return whether the type adapter uses reflection.
+   */
+  private static boolean isReflective(TypeAdapter<?> typeAdapter) {
+    if (typeAdapter instanceof ReflectiveTypeAdapterFactory.Adapter) {
+      return true;
+    } else if (typeAdapter instanceof TreeTypeAdapter) {
+      // Have to get actual serializing adapter for TreeTypeAdapter because
+      // TreeTypeAdapter without `serializer` might fall back to reflective
+      // type adapter
+      TreeTypeAdapter<?> treeTypeAdapter = (TreeTypeAdapter<?>) typeAdapter;
+      TypeAdapter<?> serializingTypeAdapter = treeTypeAdapter.getSerializingTypeAdapter();
+      if (serializingTypeAdapter == treeTypeAdapter) {
+        return false;
+      } else {
+        // Check TreeTypeAdapter delegate
+        return isReflective(serializingTypeAdapter);
+      }
+    }
+    return false;
+  }
+
   /**
    * Finds a compatible runtime type if it is more specific
    */
-  private Type getRuntimeTypeIfMoreSpecific(Type type, Object value) {
-    if (value != null
-        && (type == Object.class || type instanceof TypeVariable<?> || type instanceof Class<?>)) {
+  private static Type getRuntimeTypeIfMoreSpecific(Type type, Object value) {
+    if (value != null && (type instanceof Class<?> || type instanceof TypeVariable<?>)) {

```

Comment:
Note: Simplified this because `type instanceof Class<?>` already includes the check for `type == Object.class`.